### PR TITLE
fix: session v4

### PR DESCRIPTION
**Pull request by sattvikc**
## Summary of change

session v4 to contain `tId`

## Related issues

- https://github.com/supertokens/supertokens-core/pull/555

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your
changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here
highlighting the necessary changes)

## Checklist for important updates

- [ ] Changelog has been updated
    - [ ] If there are any db schema changes, mention those changes clearly
- [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
- [ ] `pluginInterfaceSupported.json` file has been updated (if needed)
- [ ] Changes to the version if needed
    - In `build.gradle`
- [ ] If added a new paid feature, edit the `getPaidFeatureStats` function in FeatureFlag.java file
- [ ] Had installed and ran the pre-commit hook
- [ ] If there are new dependencies that have been added in `build.gradle`, please make sure to add them
  in `implementationDependencies.json`.
- [ ] Issue this PR against the latest non released version branch.
    - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the
      latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    - If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR

- [ ] Item1
- [ ] Item2

### CODE REVIEW

File: src/main/java/io/supertokens/session/accessToken/AccessToken.java [hunk 1]
```diff
@@ -414,10 +452,10 @@ static AccessTokenInfo fromJSON(AppIdentifier appIdentifier, JsonObject payload,
                         payload.get("iat").getAsLong() * 1000,
                         version,
                         new TenantIdentifier(appIdentifier.getConnectionUriDomain(), appIdentifier.getAppId(),
-                                payload.get("tId").getAsString())
+                                payload.has("tId") ? payload.get("tId").getAsString() : null)
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/session/accessToken/AccessToken.java:455]: you should explicitly check for v3 and if it's v3, set the tenantId to null

File: src/main/java/io/supertokens/session/accessToken/AccessToken.java [hunk 2]
```diff
@@ -440,7 +478,10 @@ JsonObject toJSON() throws AccessTokenPayloadError {
                 res.addProperty("sub", this.userId);
                 res.addProperty("exp", this.expiryTime / 1000);
                 res.addProperty("iat", this.timeCreated / 1000);
-                res.addProperty("tId", this.tenantIdentifier.getTenantId());
+
+                if (this.version == VERSION.V4) {
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/session/accessToken/AccessToken.java:482]:
```suggestion
                if (this.version != VERSION.V3) {
```

Comment (rishabhpoddar) [reply src/main/java/io/supertokens/session/accessToken/AccessToken.java:482]: cause we want to do >= v4

File: src/main/java/io/supertokens/session/jwt/JWT.java [hunk 3]
```diff
@@ -90,22 +90,22 @@ public static JWTPreParseInfo preParseJWTInfo(String jwt) throws JWTException {
         }
 
         JsonPrimitive version = parsedHeader.get("version").getAsJsonPrimitive();
-        if (!version.isString() || !version.getAsString().equals("3")) {
+        if (!version.isString() || (!version.getAsString().equals("3") && !version.getAsString().equals("4"))) {
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/session/jwt/JWT.java:93]:
```suggestion
        if (!version.isString() || version.getAsString().equals("1") || version.getAsString().equals("2")) {
```

File: src/test/java/io/supertokens/test/session/api/VerifySessionAPITest2_22.java [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    @AfterClass
+    public static void afterTesting() {
+        Utils.afterTesting();
+    }
+
+    @Before
+    public void beforeEach() {
+        Utils.reset();
+    }
+
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/session/api/VerifySessionAPITest2_22.java:52]: Add tests for:
- Using v3 token with latest CDI, AccessTokenInfo points to public tenant.
- what happens when you give v4 access token to older cdi?
- what happens if you give a v2 access token to getSession -> it should give try refresh token -> refresh -> v4 token (with public tid)
- what happens if you give a v2 access token to getSession -> it should give try refresh token -> refresh (with cdi 2.21) -> v3 token (without public tid) - either update existing tests, or create new one.
- If you have tId as a custom claim in a v3 access token, getSession should succeed, the claim not change. Refresh session (with latest CDI), should override tId with public.
- If you have tId as a custom claim in a v3 access token, getSession should succeed, the claim not change. Refresh session (with CDI 2.21), should not affect tId.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/session/accessToken/AccessToken.java [hunk 5]
```diff
@@ -308,6 +309,33 @@ public static VERSION getAccessTokenVersion(AccessTokenInfo accessToken) {
...
+        }
+
+        if (version.greaterThanOrEqualTo(SemVer.v2_21)) {
+            return AccessToken.VERSION.V3;
+        }
+
+        return AccessToken.VERSION.V2;
+    }
+
+    public static VERSION getVersionFromString(String versionString) {
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/session/accessToken/AccessToken.java:324]: Done

File: src/main/java/io/supertokens/session/accessToken/AccessToken.java [hunk 6]
```diff
@@ -467,19 +508,31 @@ JsonObject toJSON() throws AccessTokenPayloadError {
...
                     throw new TryRefreshTokenException(
-                            "Access token does not contain all the information. Maybe the structure has changed?" +
+                            "Access token does not contain all the information. Maybe the structure has changed? " +
                                     prop);
                 }
             }
         }
     }
 
+    public static VERSION getLatestVersion() {
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/session/accessToken/AccessToken.java:531]: Done

File: src/main/java/io/supertokens/session/accessToken/AccessToken.java (hunk 1 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/session/accessToken/AccessToken.java:455]: Done

File: src/main/java/io/supertokens/session/accessToken/AccessToken.java (hunk 2 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/session/accessToken/AccessToken.java:482]: Done

File: src/test/java/io/supertokens/test/session/api/VerifySessionAPITest2_22.java (hunk 4 above)
Comment (sattvikc) [reply src/test/java/io/supertokens/test/session/api/VerifySessionAPITest2_22.java:58]: Using v3 token with latest CDI, AccessTokenInfo points to public tenant. --- DONE

what happens when you give v4 access token to older cdi? --- Works normally (added test)

what happens if you give a v2 access token to getSession -> it should give try refresh token -> refresh -> v4 token (with public tid) --- Doesn't force a refresh, but updates the version on refresh

what happens if you give a v2 access token to getSession -> it should give try refresh token -> refresh (with cdi 2.21) -> v3 token (without public tid) - either update existing tests, or create new one. --- similar to v4, doesn't force refresh, but updates version when refreshed. have updated existing tests

If you have tId as a custom claim in a v3 access token, getSession should succeed, the claim not change. Refresh session (with latest CDI), should override tId with public. --- verify session succeeds and also uses public tenant only and not use the tId value. on refresh, we return UNAUTHORISED stating user data contains a protected field

If you have tId as a custom claim in a v3 access token, getSession should succeed, the claim not change. Refresh session (with CDI 2.21), should not affect tId. --- works as expected

File: src/main/java/io/supertokens/session/accessToken/AccessToken.java [hunk 7]
```diff
@@ -322,22 +322,32 @@ public static VERSION getAccessTokenVersionForCDI(SemVer version) {
...
-            default:
-                throw new IllegalArgumentException("Invalid version string: " + versionString);
-        }
+        return VERSION.valueOf("V" + versionString);
     }
 
     public static class AccessTokenInfo {
-        public static String[] protectedPropNames = {
+
+        static String[] protectedPropNamesV2 = {
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/session/accessToken/AccessToken.java:330]: remove these dupluicate arrays and rename requiredPropsV2 to requiredAndProtectedPropsV2

File: src/main/java/io/supertokens/session/accessToken/AccessToken.java [hunk 8]
```diff
@@ -435,12 +445,21 @@ static AccessTokenInfo fromJSON(AppIdentifier appIdentifier, JsonObject payload,
...
                     if (Arrays.stream(protectedPropNames).noneMatch(element.getKey()::equals)) {
                         userData.add(element.getKey(), element.getValue());
                     }
                 }
 
+                TenantIdentifier tenantIdentifier;
+                if (version == VERSION.V3) {
+                    tenantIdentifier = new TenantIdentifier(appIdentifier.getConnectionUriDomain(), appIdentifier.getAppId(), null);
+                } else {
+                    tenantIdentifier = new TenantIdentifier(appIdentifier.getConnectionUriDomain(), appIdentifier.getAppId(), payload.get("tId").getAsString());
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/session/accessToken/AccessToken.java:460]: add a comment explaining why we are doing like this

File: src/main/java/io/supertokens/session/accessToken/AccessToken.java [hunk 9]
```diff
@@ -508,6 +525,15 @@ JsonObject toJSON() throws AccessTokenPayloadError {
             return res;
         }
 
+        public static String[] getProtectedProps(VERSION version) {
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/session/accessToken/AccessToken.java:528]: this function will also go away

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/session/accessToken/AccessToken.java (hunk 7 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/session/accessToken/AccessToken.java:330]: Done

File: src/main/java/io/supertokens/session/accessToken/AccessToken.java (hunk 8 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/session/accessToken/AccessToken.java:460]: Done

File: src/main/java/io/supertokens/session/accessToken/AccessToken.java (hunk 9 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/session/accessToken/AccessToken.java:528]: Done

**Pull request merged**